feat(platform): add clipboard support

WhoAteDaCake opened the pull request:
Changes proposed in this pull request:
- Add support for linux clipboard.

@lc-soft

WhoAteDaCake commented:
I've not got the text back from clipboard to the textedit widget!

The approach is still quite basic and needs more handling, but it's looking promising

WhoAteDaCake commented:
@lc-soft Could you take a look at the implementation? 

Also, I'm a bit confused why some of my commit messages don't pass `commit-lint-pr`

WhoAteDaCake commented:
Would you expect me to add a new macro called `LCUI_KEYCODE_V` and `LCUI_KEYCODE_C` ? Or convert the e->key.code to `char` and check for `'C'` or `'V'`

lc-soft commented:
> Would you expect me to add a new macro called `LCUI_KEYCODE_V` and `LCUI_KEYCODE_C` ? Or convert the e->key.code to `char` and check for `'C'` or `'V'`
> 
> Also, where can I find `LCUI_GRAPH`

Use `LCUI_KEY_V` and `LCUI_KEY_C`.

https://github.com/lc-soft/LCUI/blob/b3a0a94a9594c5c9451a2a5b0c2a8b13e5146840/include/LCUI/input.h#L66-L85

The `LCUI_Graph` type has been renamed to `pd_canvas_t`.

WhoAteDaCake commented on lib/platform/src/linux/linux_x11clipboard.c:
```diff
@@ -82,7 +82,7 @@ void ExecuteCallback(void)
 	}
 	size_t len = clipboard.text_len;
-	wchar_t *wstr = malloc(sizeof(wchar_t) * len);
+	wchar_t *wstr = malloc(sizeof(wchar_t) * (len + 1));
```
Is there a need for the `+1` here? I'm not quite sure if I'm doing the copying right. Would it be possible to extract this to a macro or a function, so these mistakes can be avoided?
@lc-soft

lc-soft replied:
@WhoAteDaCake  Yes. the string terminator `'\0'` also takes up memory, so we need `len + 1`.

lc-soft replied:
When decoding a string, we need to get the length of the decoded string before allocating memory.

```c
size_t len = decode_utf8(NULL, clipboard.text, 0);
wchar_t *wstr = malloc(sizeof(wchar_t) * (len + 1));
decode_utf8(wstr, clipboard.text, len);
wstr[len] = 0;
```

lc-soft commented on lib/platform/src/linux/linux_x11clipboard.c:
```diff
@@ -82,7 +82,7 @@ void ExecuteCallback(void)
-	wchar_t *wstr = malloc(sizeof(wchar_t) * len);
+	wchar_t *wstr = malloc(sizeof(wchar_t) * (len + 1));
 	decode_utf8(wstr, clipboard.text, len);
 	wstr[len + 1] = 0;
```
```diff
- 	wstr[len + 1] = 0;
+ 	wstr[len] = 0;
```

WhoAteDaCake commented:
@lc-soft I'm trying to start testing the code, however, I'm encountering issues while trying to add is as taget:

```
test clipboard
error: execv(../build/linux/x86_64/debug/run-tests) failed(-1)
warning: cannot match target(run-tests).add_files("test/run_tests.c") at ./test/xmake.lua:10
warning: cannot match target(run-tests).add_files("test/cases/*.c") at ./test/xmake.lua:10
```

Any tips?

WhoAteDaCake commented:
I've added a copy paste test, does it look alright? Is there a need add additional tests?

lc-soft commented:
> @lc-soft I'm trying to start testing the code, however, I'm encountering some warnings:
> 
> ```
> test clipboard
> warning: cannot match target(run-tests).add_files("test/run_tests.c") at ./test/xmake.lua:10
> warning: cannot match target(run-tests).add_files("test/cases/*.c") at ./test/xmake.lua:10
> ```
> 
> Any tips?

In which directory did you run `xmake run run tests`?

lc-soft commented on test/cases/test_clipboard.c:
```diff
@@ -0,0 +1,86 @@
+	TextEdit_SetTextW(text1, L"helloworld");
+	LCUIWidget_SetFocus(text1);
+
+	LCUIThread_Create(&tid, ObserverThread, &exited);
```
ObserverThread is not required. Please delete `exited` `tid` and `ObserverThread`.

lc-soft commented on test/run_tests.c:
```diff
@@ -10,21 +10,22 @@ int main(void)
+	// describe("test image reader", test_image_reader);
+	// describe("test xml parser", test_xml_parser);
+	// describe("test widget event", test_widget_event);
+	// describe("test widget opacity", test_widget_opacity);
```
These changes should be cancelled

lc-soft commented on test/run_tests.h:
```diff
@@ -19,3 +19,4 @@ void test_mainloop(void);
 void test_block_layout(void);
 void test_flex_layout(void);
 void test_widget_rect(void);
+void test_clipboard(void);
```
A newline is required at the end of file.

![image](https://user-images.githubusercontent.com/1730073/141685033-8ac7a128-c542-4298-8fad-c6a85af397c4.png)

WhoAteDaCake commented:
> > @lc-soft I'm trying to start testing the code, however, I'm encountering some warnings:
> > ```
> > test clipboard
> > warning: cannot match target(run-tests).add_files("test/run_tests.c") at ./test/xmake.lua:10
> > warning: cannot match target(run-tests).add_files("test/cases/*.c") at ./test/xmake.lua:10
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any tips?
> 
> In which directory did you run `xmake run run tests`?

I've ran it from the project run directory.

WhoAteDaCake commented:
Thank you for your patience

lc-soft commented:
`Run tests`  took too long

![image](https://user-images.githubusercontent.com/1730073/141713679-a28e1fcd-7d23-4b90-bddc-cbe7652de4ae.png)

WhoAteDaCake commented:
Could it be because the test requires x11 clipboard to be available within the running machine and it's not there?

lc-soft commented:
> Could it be because the test requires x11 clipboard to be available within the running machine and it's not there? That's the reason, why I've added that timeout Observer thread, I wasn't if it would be caught if something like this had happened.
> 
> Edit: No it seems like it didn't get to that point, I'll investigate this later today

I think you should add observer thread to the `linux_x11clipboard.c` file and pass empty data to the callback when the clipboard request times out.

lc-soft commented:
I recommend that you add a default clipboard driver for use when X11 and Windows clipboard drivers are not available.

WhoAteDaCake commented:
> I think you should add observer thread to the linux_x11clipboard.c file and pass empty data to the callback when the clipboard request times out.

Should we maybe indicate if it failed in some sort of way as well? Since if empty text is returned, there is no way to know whether nothing is copied or it just failed.

> I recommend that you add a default clipboard driver for use when X11 and Windows clipboard drivers are not available.
Where should i place this? Create a new folder in `/lib/platform/src/` called `headless` or just place `clipboard.c` there?

lc-soft commented:
> Should we maybe indicate if it failed in some sort of way as well? Since if empty text is returned, there is no way to know whether nothing is copied or it just failed.

Pass NULL to the callback to tell it that the clipboard request failed.

```c
callback->action(NULL, callback->arg);
```

> Where should i place this? Create a new folder in /lib/platform/src/ called headless or just place clipboard.c in /lib/platform/src/?

Please ignore my last comment because it's a bit complicated to implement. 

An easier way to fix build failures in Windows is to add functions such as `LCUI_UseClipboard` without their implementation code.

```c
void LCUI_UseClipboard(LCUI_ClipboardAction action, void *arg)
{
    // Empty
}
```

WhoAteDaCake commented:
I've added an observer thread, in regards to testing the clipboard, how should I go about testing it? Thanks.

lc-soft commented:
> I've added an observer thread, in regards to testing the clipboard, how should I go about testing it? Thanks.


I think the test method in the current `test_clipboard.c` file is ok.

WhoAteDaCake commented:
But since there is no clipboard access in the CI pipeline, it won't work right? All we'd get is a timeout

WhoAteDaCake commented:
@lc-soft

So the reason it hangs it's because inside `linux_clipboard.c` `LCUI_UseClipboard` implementation is empty when X11 is not used, so the `"change"` event within the test never gets triggered.

I could try and add an In file clipboard for Linux within the `linux_clipboard.c`, which would be used when x11 is not available.

lc-soft commented on lib/platform/src/linux/linux_x11clipboard.c:
```diff
@@ -0,0 +1,337 @@
+{
+	// X11 doesn't support wchar_t, so we need to send it regular char
+	char *raw_text = malloc((len + 1) * sizeof(char));
+	size_t raw_len = wcstombs(raw_text, text, len);
```
I'm not sure `wchar_t` is a better clipboard text type than `char`, but after seeing this line of code I think we can use char `instead` of `wchar_t` so that the clipboard implementation doesn't have to do the string encoding and decoding, what do you think?

WhoAteDaCake replied:
If windows clipboard needs char as well, then I think we should convert at `textedit` level, if it supports wchar_t, then for the sake of consistence I'd keep it as it is

lc-soft commented:
> @lc-soft
> 
> So the reason it hangs it's because inside `linux_clipboard.c` `LCUI_UseClipboard` implementation is empty when X11 is not used, so the `"change"` event within the test never gets triggered.
> 
> I could try and add an In file clipboard for Linux within the `linux_clipboard.c`, which would be used when x11 is not available.

Try binding the `"paste"` event handler

lc-soft commented:
> But since there is no clipboard access in the CI pipeline, it won't work right? As we'd get is a timeout for when we request to own the clipboard thinking . I'll investigate this further

Yes, I will test this pull request with WSL in a few days.

WhoAteDaCake commented:
> Try binding the "paste" event handler

The paste event won't get called either, since `LCUI_UseClipboard` never sets the callback.

lc-soft commented on lib/ui-widgets/src/textedit.c:
```diff
@@ -714,6 +717,23 @@ static void TextEdit_TextDelete(LCUI_Widget widget, int n_ch)
+static void TextEdit_OnPaste(LCUI_Widget w, LCUI_WidgetEvent e, void *arg)
+{
+	LCUI_Clipboard text = arg;
+	TextEdit_InsertTextW(w, text->text);
```
```c
	LCUI_Clipboard clipboard = arg;
	if (clipboard) {
		TextEdit_InsertTextW(w, clipboard->text);
	}
```

lc-soft commented on test/cases/test_clipboard.c:
```diff
@@ -0,0 +1,64 @@
+	LCUI_TriggerEvent(&ev, NULL);
+}
+
+static void OnCheckText(LCUI_Widget w, LCUI_WidgetEvent e, void *arg)
```
```diff
+ #include <LCUI/clipboard.h>
```
```c
static void OnCheckText(LCUI_Widget w, LCUI_WidgetEvent e, void *arg)
{
	LCUI_Clipboard clipboard = arg;
	it_b("check the pasted text",
	     clipboard != NULL && wcscmp(clipboard->text, L"helloworld") == 0, TRUE);
	LCUI_Quit();
}
```

WhoAteDaCake replied:
That still won't work right now, since just a NULL is passed without any text. How about, I create a storage for clipboard within `linux_clipboard.c` ? Since it can act as a dummy way to test things.

Also, nice to see you back @lc-soft

WhoAteDaCake commented on lib/platform/src/linux/linux_x11clipboard.c:
```diff
@@ -0,0 +1,337 @@
+	char *raw_text = malloc((len + 1) * sizeof(char));
+	size_t raw_len = wcstombs(raw_text, text, len);
+	raw_text[raw_len] = '\0';
+	free(text);
```
So free should happen in `TextEdit_OnKeyDown`

WhoAteDaCake replied:
This means the free should happen within `TextEdit_OnKeyDown` right? After the `LCUI_SetClipboardText`

WhoAteDaCake commented:
Does that work @lc-soft ?

WhoAteDaCake commented:
I think it would be useful to have this kind of conversion between (char -> w_char) and (w_char -> char) extracted as a utility function, since it's easy to mess up, and it can be useful in many places

lc-soft commented:
> Does that work @lc-soft ?

Please try to replace `#include "config.h"` with `#include "../config.h"` in directory `lib/platform/src/linux/`

![image](https://user-images.githubusercontent.com/1730073/151663941-7ef7b684-4416-4e9c-ba85-f759480bb7f5.png)

WhoAteDaCake commented:
> > Does that work @lc-soft ?
> 
> Please try to replace `#include "config.h"` with `#include "../config.h"` in directory `lib/platform/src/linux/`
> 
> ![image](https://user-images.githubusercontent.com/1730073/151663941-7ef7b684-4416-4e9c-ba85-f759480bb7f5.png)

I think the failing tests aren't related to the config right?

WhoAteDaCake commented:
```
  root width 320px
    √ $('.example')[0].box.border == (10, 10, 286, 604)
    √ $('.example')[1].box.border == (10, 624, 286, 304)
    × $('.example')[2].box.border == (10, 938, 286, 280)
      AssertionError: (10, 938, 286, 242) == (10, 938, 286, 280)
      + expected - actual

      - (10, 938, 286, 242)
      + (10, 938, 286, 280)

    × $('.example')[3].box.border == (10, 1228, 286, 732)
      AssertionError: (10, 1190, 286, 732) == (10, 1228, 286, 732)
      + expected - actual

      - (10, 1190, 286, 732)
      + (10, 1228, 286, 732)

    × $('.example')[4].box.border == (10, 1970, 286, 504)
      AssertionError: (10, 1932, 286, 504) == (10, 1970, 286, 504)
      + expected - actual

      - (10, 1932, 286, 504)
      + (10, 1970, 286, 504)

    × $('.example')[5].box.border == (10, 2484, 286, 254)
      AssertionError: (10, 2446, 286, 254) == (10, 2484, 286, 254)
      + expected - actual

      - (10, 2446, 286, 254)
      + (10, 2484, 286, 254)
```

lc-soft commented:
> ```
>   root width 320px
>     √ $('.example')[0].box.border == (10, 10, 286, 604)
>     √ $('.example')[1].box.border == (10, 624, 286, 304)
>     × $('.example')[2].box.border == (10, 938, 286, 280)
>       AssertionError: (10, 938, 286, 242) == (10, 938, 286, 280)
>       + expected - actual
> 
>       - (10, 938, 286, 242)
>       + (10, 938, 286, 280)
> 
>     × $('.example')[3].box.border == (10, 1228, 286, 732)
>       AssertionError: (10, 1190, 286, 732) == (10, 1228, 286, 732)
>       + expected - actual
> 
>       - (10, 1190, 286, 732)
>       + (10, 1228, 286, 732)
> 
>     × $('.example')[4].box.border == (10, 1970, 286, 504)
>       AssertionError: (10, 1932, 286, 504) == (10, 1970, 286, 504)
>       + expected - actual
> 
>       - (10, 1932, 286, 504)
>       + (10, 1970, 286, 504)
> 
>     × $('.example')[5].box.border == (10, 2484, 286, 254)
>       AssertionError: (10, 2446, 286, 254) == (10, 2484, 286, 254)
>       + expected - actual
> 
>       - (10, 2446, 286, 254)
>       + (10, 2484, 286, 254)
> ```

This problem can be ignored because the test program loaded different fonts in the test environment than in my system, and rendered text with these fonts was smaller, resulting in a layout that was not as expected.

lc-soft commented on test/run_tests.c:
```diff
@@ -26,5 +26,6 @@ int main(void)
 	describe("test block layout", test_block_layout);
 	describe("test flex layout", test_flex_layout);
 	describe("test widget rect", test_widget_rect);
+	// describe("test clipboard", test_clipboard);
```
```diff
- // describe("test clipboard", test_clipboard);
+ describe("test clipboard", test_clipboard);
```

WhoAteDaCake replied:
Yes, I'm just trying to figure out here why the test is hanging

lc-soft reviewed (changes requested):
I plan to merge this PR as soon as possible, because my other PR conflicts with it, and I will deal with its remaining problems in another PR. Please confirm that you have pushed all the changes related to this PR.

WhoAteDaCake commented:
Doing the last check of trying to figure out why the test is failing and then I can confirm

WhoAteDaCake commented:
Just as a final note. Is `#USE_LIBX11` always set? Even in CI environment it seems to be set, that's why after I added local data to `linux_clipboard.c` it hung because I added `else` to the macro.

WhoAteDaCake commented:
It's a tiny issue though, the clipboard tests are now passing, it's a small start, but hopefully we can improve it in the future, don't want to delay your other work too much!

WhoAteDaCake commented:
Or potentially, I can store that data and free the previous allocation on the next ask for clipboard, this way there is no need to try and communicate if multiple listeners are created for the paste event, what do you think?

lc-soft commented:
> Just as a final note. Is `#USE_LIBX11` always set? Even in CI environment it seems to be set, that's why after I added local data to `linux_clipboard.c` it hung because I added `else` to the macro.

Yes, it is, but you should also handle `#USE_LIBX11` when it is not set.

![image](https://user-images.githubusercontent.com/1730073/151737113-b838974a-207e-417f-8e1c-31f6988d24be.png)

lc-soft commented:
> Last fix: there is a memory leak, since after clipboard text + `clipboard_data` is sent, it's not freed. Which part of code should free it? `textedit` or `clipboard`

`clipboard`

lc-soft commented on lib/platform/src/linux/linux_clipboard.c:
```diff
@@ -0,0 +1,109 @@
+		len = decode_utf8(wstr, clipboard.text, len);
+		wstr[len] = 0;
+		// Assign the data
+		LCUI_Clipboard clipboard_data = malloc(sizeof(LCUI_ClipboardRec));
```
```diff
- LCUI_Clipboard clipboard_data = malloc(sizeof(LCUI_ClipboardRec));
+ LCUI_ClipboardRec clipboard_data = { 0 };
```
```diff
- action(clipboard_data, arg);
+ action(&clipboard_data, arg);
```

WhoAteDaCake replied:
Can we assume that the action always happens synchronously? So as soon as action is done the string is no longer used.

lc-soft replied:
> Can we assume that the action always happens synchronously? So as soon as action is done the string is no longer used.

Yes, we can.

WhoAteDaCake commented:
> > Just as a final note. Is `#USE_LIBX11` always set? Even in CI environment it seems to be set, that's why after I added local data to `linux_clipboard.c` it hung because I added `else` to the macro.
> 
> Yes, it is, but you should handle the case where `#USE_LIBX11` is not set.
> 
> ![image](https://user-images.githubusercontent.com/1730073/151737113-b838974a-207e-417f-8e1c-31f6988d24be.png)

In that case, we should not set it in environments like CI, as the handling would be exactly the same as if `LCUI_GetAppId() != LCUI_APP_LINUX_X11`

WhoAteDaCake commented:
Or I can just extract it to a function and make sure both variants call it

WhoAteDaCake commented:
The tests are now passing

WhoAteDaCake commented:
Thank you for all the guidance!

lc-soft commented:
@WhoAteDaCake I made some changes to make the clipboard tests work on Windows.

https://github.com/lc-soft/LCUI/blob/v3.0-dev/lib/platform/src/clipboard.c

![image](https://user-images.githubusercontent.com/1730073/152111117-859713a2-31d8-4e59-beb7-e180cbdc5440.png)
